File: Typoceros/Typo.py
# Get the path of the script
from dataclasses import dataclass, field
from math import log2
from typing import Callable, Generator, List, Optional, Tuple

import regex as re

from .util import *


@dataclass
class Typo:
   """Class for Typo Engine."""

   text: str = field(repr=False)
   _length: int = field(init=False, repr=False)
   _slots: Optional[List[Tuple[Tuple[int, int], str, re.Pattern]]] = field(init=True, repr=False,default=None)
   _spaces: Optional[List[int]] = field(init=True, repr=False,default=None)
   verbose: bool = field(init=True,repr=False,default=False)

   def __post_init__(self):
      if self.text != normalize(self.text,self.verbose):
         raise ValueError("Text isn't spelled correctly")
   @staticmethod
   def isAcceptable(text:str,verbose:bool=False):
      return text == normalize(text,verbose)
   @staticmethod
   def FixText(text:str,verbose=False):
      return normalize(text,verbose)
   def apply(self, space: int, offset: int, text: str) -> str:
      if self.verbose:
         print(f"apply: space={space}, offset={offset}, text={text}")
      if offset == 0:
         return text
      match_tuple = self.slots[sum(self.spaces[0:space]) + offset - 1]
      applied = apply_match(text, match_tuple,self.verbose)
      if self.verbose:
         print(f"applied: {applied}")
      return applied
   @property
   def slots(self):
      if self._slots is None:
         self._slots = valid_rules_scan(self.text,self.verbose)
      return self._slots
   @property
   def length(self) -> int:
      return len(self.slots)
   @length.setter
   def length(self, length: int):
      pass
   @property
   def spaces(self) -> List[int]:
      if self._spaces is not None:
         return self._spaces
      
      sentence_ranges = chunker(self.text)
      
      # Initialize an empty list of buckets
      num_buckets = len(sentence_ranges)
      buckets: List[int] = [0 for _ in range(num_buckets)]
      
      # Iterate through each element range and put it in the corresponding bucket
      for i, (start, end) in enumerate(span for span,_,_ in self.slots):
         for j, (sent_start, sent_end) in enumerate(sentence_ranges):
            if sent_start <= start < sent_end and sent_start < end <= sent_end:
               buckets[j] += 1
               break
      return buckets
   @spaces.setter
   def spaces(self, value):
      pass
   @property
   def bits(self):
      return list(map(int, map(lambda x : log2(x + 1), self.spaces)))
   @bits.setter
   def bits(self, bits: int):
      pass
   def encode(self, values:List[int]):
      spaces = self.spaces
      if len(values) > len(spaces):
         raise ValueError("Can't encode")
      for i in range(len(values)):
         # spaces[i] = 0 means that the chunk has a birth defect
         # a typo not by us making the chunk unusable and in that case 
         # values[i] = 0 and the fact that it's an un-fixable typo will
         # tell the decoder to learn it
         if values[i] >= spaces[i] and spaces[i] != 0:
            raise ValueError("Won't fit")
      result = self.text
      for i in range(len(values) - 1, -1, -1):
         result = self.apply(i, values[i], result)
      return result
   @staticmethod
   def decode(text:str,verbose=False,test_self=None) -> Tuple[str,List[int]]:
      original = normalize(text,verbose)
      if test_self is not None:
         if original != test_self.text:
            print(f'original=\n{original}')
            print(f'test_self.text=\n{test_self.text}')
         assert original == test_self.text
      t = Typo(original)

      return original, t._decode(text,test_self)
   def _decode(self, text:str,test=None) -> List[int]:
      a_self = test if test is not None else self
      spaces = a_self.spaces
      cnt = len(diff(text,a_self.text))
      if a_self.verbose:
         print(f'cnt={cnt}')
         print(f'diff(text,a_self.text)={diff(text,a_self.text)}')
      values = [0 for s in spaces]
      for index, space in enumerate(spaces):
         isZero = True
         for i in range(space):
            values[index] = i
            dif = diff(text, a_self.encode(values))
            if len(dif) == cnt - 1:
               if a_self.verbose:
                  print(f'values={values}')
                  print(f'dif={dif}')
               cnt -= 1
               isZero = False
               break 
         if isZero:
            values[index] = 0     
            if a_self.verbose:
               print(f'chunk is empty values={values}')
      return  values
   def encode_encoder(self, bytes_str: str) -> Tuple[List[int], str]:
      if not set(bytes_str) <= set('01'):
         raise ValueError(f"bytes_str isn't a bytes string : '{bytes_str}'")
      values = self.bits
      bit_values = []
      remaining_bits = bytes_str
      for i, val in enumerate(values):
         if len(remaining_bits) >= val + 1 and int(remaining_bits[:val+1]) < self.spaces[i]:
            bit_value = int(remaining_bits[:val+1], 2)
            bit_values.append(bit_value)
            remaining_bits = remaining_bits[val+1:]
         elif len(remaining_bits) >= val and val > 0:
            bit_value = int(remaining_bits[:val], 2)
            bit_values.append(bit_value)
            remaining_bits = remaining_bits[val:]
         else:
            bit_values.append(0)
      return bit_values, remaining_bits
   def learn(self,text:str)->None:
      normalize(text,learn=True,verbose=self.verbose)

File: Typoceros4j/src/main/java/Typo.java
import com.google.common.collect.Lists;
import io.vavr.Tuple2;
import io.vavr.Tuple3;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;
import java.util.regex.Pattern;


public class Typo {
    private String text;
    private int _length;
    private List<Tuple3<Span, String, Pattern>> _slots = null;
    private List<Integer> _spaces = null;
    private boolean verbose;
    public int span_size = 6;

    public int getLength() throws IOException {
        return getSlots().size();
    }

    public List<Tuple3<Span, String, Pattern>> getSlots() throws IOException {
        if (this._slots == null)
            this._slots = LangProxy.valid_rules_scan(this.text, this.verbose);
        return this._slots;
    }

    public List<Integer> getSpaces() throws IOException {
        if (_spaces != null) {
            return _spaces;
        }

        var sentenceRanges = util.chunk(text, this.span_size);

        // Initialize an empty list of buckets
        int numBuckets = sentenceRanges.size();
        List<Integer> buckets = new ArrayList<>(Arrays.asList(new Integer[numBuckets]));
        Collections.fill(buckets, 0);

        // Iterate through each element range and put it in the corresponding bucket
        for (int i = 0; i < this.getSlots().size(); i++) {
            var slot = this.getSlots().get(i)._1;
            int start = slot.start;
            int end = slot.end;
            for (int j = 0; j < sentenceRanges.size(); j++) {
                var sentenceRange = sentenceRanges.get(j);
                int sentStart = sentenceRange.start;
                int sentEnd = sentenceRange.end;
                if (sentStart <= start && start < sentEnd && sentStart < end && end <= sentEnd) {
                    buckets.set(j, buckets.get(j) + 1);
                    break;
                }
            }
        }
        _spaces = buckets;
        return buckets;
    }

    public List<Integer> getBits() throws IOException {
        return getSpaces().stream().map(util::log2).collect(Collectors.toList());
    }

    public Typo(String text) throws IOException {
        _constructor(text, false);
    }

    public Typo(String text, boolean verbose) throws IOException {
        _constructor(text, verbose);
    }

    public void _constructor(String text, boolean verbose) throws IOException {
        this.text = text;
        this.verbose = verbose;
        if (!isAcceptable(this.text, this.verbose)) {
            throw new IllegalArgumentException("Text isn't spelled correctly");
        }
    }

    public boolean isAcceptable(String text, boolean verbose) throws IOException {
        return text.equals(LangProxy.normalize(text, verbose));
    }

    public String FixText(String text, boolean verbose) throws IOException {
        return LangProxy.normalize(text, verbose);
    }

    public String apply(int space, int offset, String text) throws IOException {
        if (this.verbose) {
            System.out.println("apply: space=" + space + ", offset=" + offset + ", text=" + text);
        }
        if (offset == 0) {
            return text;
        }
        var match_tuple = this.getSlots()
                .get(util.sum(this.getSpaces().subList(0, space)) + offset - 1);
        var applied = LangProxy.applyMatch(text, match_tuple, this.verbose);
        if (this.verbose) {
            System.out.println("applied: " + applied);
        }
        return applied;
    }

    public String encode(List<Integer> values) throws IllegalArgumentException, IOException {
        var spaces = this.getSpaces();
        if (values.size() > spaces.size()) {
            throw new IllegalArgumentException("Can't encode");
        }

        for (int i = 0; i < values.size(); i++) {
            if (values.get(i) >= spaces.get(i) && spaces.get(i) != 0) {
                throw new IllegalArgumentException("Won't fit");
            }
        }

        String result = this.text;
        for (int i = values.size() - 1; i >= 0; i--) {
            result = this.apply(i, values.get(i), result);
        }

        return result;
    }

    public static Tuple2<String, List<Integer>> decode(String text, boolean verbose, Typo test_self) throws IOException {
        String original = LangProxy.normalize(text, verbose);
        if (verbose)
            System.out.println("original=" + original);
        Typo t;
        if (test_self != null) {
            if (!original.equals(test_self.text)) {
                System.out.println("test_self.text=\n" + test_self.text);
            }
            assert original.equals(test_self.text);
            t = test_self;
        } else {
            t = new Typo(original, verbose);
        }
        return new Tuple2<>(original, t._decode(text, test_self));
    }

    public List<Integer> _decode(String text, Typo test) throws IOException {
        Typo a_self = test != null ? test : this;
        var spaces = a_self.getSpaces();
        int cnt = util.diff(text, a_self.text).size();
        if (a_self.verbose) {
            System.out.println("cnt=" + cnt);
            System.out.println("util.diff('" + text + "','" + a_self.text + "')=" + util.diff(text, a_self.text));
        }
        List<Integer> values = new ArrayList<>(Collections.nCopies(spaces.size(), 0));
        for (int i = 0; i < spaces.size(); i++) {
            for (int j = 0; j < spaces.get(i); j++) {
                var dif = util.diff(text, a_self.encode(values));
                if (dif.size() == cnt - 1) {
                    if (a_self.verbose) {
                        System.out.println("values=" + values);
                        System.out.println("dif=" + dif);
                    }
                    cnt--;
                    break;
                }
                values.set(i, j + 1);
            }
            if (Objects.equals(values.get(i), spaces.get(i))) {
                values.set(i, 0);
                if (a_self.verbose) {
                    System.out.println("chunk is empty values=" + values);
                }
            }
        }
        return values;
    }


    public Tuple2<List<Integer>, String> encode_encoder(String bytes_str) throws ValueError, IOException {
        if (!new HashSet<>(Arrays.asList('0', '1')).containsAll(new HashSet<>(bytes_str.chars().mapToObj(c -> (char) c).collect(Collectors.toList())))) {
            throw new ValueError("bytes_str isn't a bytes string : '" + bytes_str + "'");
        }
        List<Integer> values = this.getBits();
        List<Integer> bit_values = new ArrayList<>();
        String remaining_bits = bytes_str;
        for (int i = 0; i < values.size(); i++) {
            int val = values.get(i);
            if (remaining_bits.length() >= val + 1 &&
                    Integer.parseInt(remaining_bits.substring(0, val + 1), 2) < this.getSpaces().get(i)) {
                int bit_value = Integer.parseInt(remaining_bits.substring(0, val + 1), 2);
                bit_values.add(bit_value);
                remaining_bits = remaining_bits.substring(val + 1);
            } else if (remaining_bits.length() >= val && val > 0) {
                int bit_value = Integer.parseInt(remaining_bits.substring(0, val), 2);
                bit_values.add(bit_value);
                remaining_bits = remaining_bits.substring(val);
            } else {
                bit_values.add(0);
            }
        }
        return new Tuple2<>(bit_values, remaining_bits);
    }

    public void learn(String text) throws IOException {
        LangProxy.normalize(text, this.verbose, true);
    }
}